main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1077 @@
+    case UnkindedType.Constructor.Float32 => Kind.Star
+    case UnkindedType.Constructor.Float64 => Kind.Star
+    case UnkindedType.Constructor.Int8 => Kind.Star
+    case UnkindedType.Constructor.Int16 => Kind.Star
+    case UnkindedType.Constructor.Int32 => Kind.Star
+    case UnkindedType.Constructor.Int64 => Kind.Star
+    case UnkindedType.Constructor.BigInt => Kind.Star
+    case UnkindedType.Constructor.Str => Kind.Star
+    case UnkindedType.Constructor.Arrow(arity) => Kind.Bool ->: Kind.mkArrow(arity)
+    case UnkindedType.Constructor.RecordEmpty => Kind.Record
+    case UnkindedType.Constructor.RecordExtend(field) => Kind.Star ->: Kind.Record ->: Kind.Record
+    case UnkindedType.Constructor.SchemaEmpty => Kind.Schema
+    case UnkindedType.Constructor.SchemaExtend(pred) => Kind.Star ->: Kind.Schema ->: Kind.Schema
+    case UnkindedType.Constructor.Array => Kind.Star ->: Kind.Star
+    case UnkindedType.Constructor.Channel => Kind.Star ->: Kind.Star
+    case UnkindedType.Constructor.Lazy => Kind.Star ->: Kind.Star
+    case UnkindedType.Constructor.Tag(sym, tag) => Kind.Star ->: Kind.Star ->: Kind.Star
+    case UnkindedType.Constructor.Enum(sym) => getEnumKind(root.enums(sym))
+    case UnkindedType.Constructor.Native(clazz) => Kind.Star
+    case UnkindedType.Constructor.ScopedRef => Kind.Star ->: Kind.Bool ->: Kind.Star
+    case UnkindedType.Constructor.Tuple(l) => Kind.mkArrow(l)
+    case UnkindedType.Constructor.Relation => Kind.Star ->: Kind.Star
+    case UnkindedType.Constructor.Lattice => Kind.Star ->: Kind.Star
+    case UnkindedType.Constructor.True => Kind.Bool
+    case UnkindedType.Constructor.False => Kind.Bool
+    case UnkindedType.Constructor.Not => Kind.Bool ->: Kind.Bool
+    case UnkindedType.Constructor.And => Kind.Bool ->: Kind.Bool ->: Kind.Bool
+    case UnkindedType.Constructor.Or => Kind.Bool ->: Kind.Bool ->: Kind.Bool
+    case UnkindedType.Constructor.Region => Kind.Bool ->: Kind.Star
+  }
+
+  /**
+    * Describes a kind bound, a subkind or superkind of some kind.
+    */
+  private case class KindMatch(assoc: KindMatch.Association, kind: Kind) {
+    /**
+      * Returns true if the given kind obeys the kind boudn described by this KindMatch.
+      */
+    def matches(other: Kind): Boolean = assoc match {
+      case KindMatch.Association.SubKind => other <:: kind
+      case KindMatch.Association.SuperKind => kind <:: other
+    }
+  }
+
+  private object KindMatch {
+
+
+    /**
+      * A KindMatch that matches every kind.
+      */
+    // association doesn't matter
+    val wild: KindMatch = KindMatch(Association.SubKind, Kind.Wild)
+
+    def subKindOf(kind: Kind): KindMatch = KindMatch(Association.SubKind, kind)
+
+    def superKindOf(kind: Kind): KindMatch = KindMatch(Association.SuperKind, kind)
+
+    /**
+      * Describes whether the KindMatch is a lower or upper bound.
+      */
+    sealed trait Association
+
+    object Association {
+      case object SubKind extends Association
```

Comment:
So this is because of our incorrect representation of records/schemas?

Reply:
Yes, since there ends up being contravariance in the kinding of type lambdas, we have to be able to check both subkinding and superkinding.